src/message.rs

```diff
@@ -11,20 +12,58 @@ use crate::escaped::{escape_char, unescape_char};
 
 #[derive(Debug, PartialEq, Default)]
 pub struct Message<'a> {
-    pub tags: BTreeMap<&'a str, Cow<'a, str>>,
-    pub prefix: Option<&'a str>,
-    pub command: &'a str,
-    pub params: Vec<&'a str>,
+    pub tags: BTreeMap<Cow<'a, str>, Cow<'a, str>>,
+    pub prefix: Option<Cow<'a, str>>,
+    pub command: Cow<'a, str>,
+    pub params: Vec<Cow<'a, str>>,
 }
 
-fn parse_tags<'a>(input: &'a str) -> Result<BTreeMap<&'a str, Cow<'a, str>>, Error> {
+impl<'a> Message<'a> {
+    pub fn into_owned(self) -> Message<'static> {
+        Message {
+            tags: BTreeMap::from_iter(
+                self.tags
+                    .into_iter()
+                    .map(|(k, v)| (Cow::Owned(k.into_owned()), Cow::Owned(v.into_owned()))),
+            ),
+            prefix: self.prefix.map(|s| Cow::Owned(s.into_owned())),
+            command: Cow::Owned(self.command.to_string()),
+            params: self
+                .params
+                .into_iter()
+                .map(|s| Cow::Owned(s.into_owned()))
+                .collect(),
+        }
+    }
+
+    pub fn to_owned(&self) -> Message<'static> {
+        Message {
+            tags: BTreeMap::from_iter(
+                self.tags
+                    .iter()
+                    .map(|(k, v)| (Cow::Owned(k.to_string()), Cow::Owned(v.to_string()))),
+            ),
+            prefix: self.prefix.as_ref().map(|s| Cow::Owned(s.to_string())),
+            command: Cow::Owned(self.command.to_string()),
+            params: self
+                .params
+                .iter()
+                .map(|s| Cow::Owned(s.to_string()))
+                .collect(),
+        }
+    }
+}
```

Comment:
I don't think you need two functions:

The implementation of ToString for Cow<'_, str> just calls to_owned(), and Cow uses the default implementation of to_owned(), it looks like.  So either function is effectively the same as just cloning the Cow.  Using into_owned() is a bit better, since it will only clone the Cow if it is Borrowed.  If it is Owned, then into_owned() simply returns the Cow back to you.

So, I'd recommend cutting this down to a single function that uses into_owned() under the hood instead of to_string().  **BUT**, I would recommend making that single function a to_owned() function.  It's actually a standard trait (ToOwned), so just put it in an impl block for that trait.

Cow's into_owned() is a bit of an oddball function; I think it only exists because Cow needs to be a little weird with ownership.  It has the ToOwned trait, which it just uses the aforementioned default implementation for.  But it needed to have a second, similar function given it's intended clone-on-write functionality, which is where into_owned() came from (which does not come from any sort of trait).  Since you're not implementing a special clone-on-write container, I'd recommend avoiding into_owned() and just sticking with a custom implementation of the ToOwned trait.  Plus, if you want to be able to keep the ability to just clone a Message, you can just derive the Clone trait instead.

Reply:
Here's the problem: you can't implement ToOwned for this type because you can't implement Borrow.

Also into_owned takes the value, to_owned takes a reference. That's the main difference.

Reply:
I'm pretty sure both return values.  Cow uses the blanket implementation of to_owned(), which just returns a clone of itself (and not as a ref): https://doc.rust-lang.org/src/alloc/borrow.rs.html#85-87

I'm realizing that looking at Cow's implementation of to_owned() isn't actually helpful here, however (I just realized that to_string() is actually calling to_owned() on the underlying str, not the Cow itself).  So that's my bad.  But, the main point is that you're still going to be cloning the underlying str every time you call to_string(), and then sticking that in a new Cow::Owned.

into_owned() also returns a value, and is almost identical except for the fact that it simply returns itself if it's already Owned.  https://doc.rust-lang.org/src/alloc/borrow.rs.html#313-318.

So, as far as I can tell, it looks like both functions do basically the same thing: take a Message, take all the strings out of the Cows in the message and stick them in new Owned Cows, and then put those new Cows in a new message.  The only difference is that into_owned will simply move the value instead of cloning it if it's already Owned (which seems preferable to cloning every time, _especially_ if you don't want to effectively invalidate any other Borrowed Cows that are looking at the same values).


I do need to take back what I said about simply deriving Clone, however.  This will result in clone being called directly on the Cows, _not_ on the underlying data.  So if the Cow is already Borrowed, then the new clone will also be Borrowed  (unless that's the behavior you want).